sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSinkLog.scala

```diff
@@ -0,0 +1,241 @@
+    }
+    val version = lines(0)
+    if (version != VERSION) {
+      throw new IllegalStateException(s"Unknown log version: ${version}")
+    }
+    lines.toSeq.slice(1, lines.length).map(read[FileLog](_))
+  }
+
+  override def add(batchId: Long, logs: Seq[FileLog]): Boolean = {
+    if (isCompactionBatch(batchId, compactLength)) {
+      compact(batchId, logs)
+    } else {
+      super.add(batchId, logs)
+    }
+  }
+
+  /**
+   * Compacts all logs before `batchId` plus the provided `logs`, and writes them into the
+   * corresponding `batchId` file.
+   */
+  private def compact(batchId: Long, logs: Seq[FileLog]): Boolean = {
+    val validBatches = getValidBatchesBeforeCompactionBatch(batchId, compactLength)
+    val allLogs = validBatches.flatMap(batchId => get(batchId)).flatten ++ logs
+    if (super.add(batchId, compactLogs(allLogs))) {
+      if (isDeletingExpiredLog) {
+        deleteExpiredLog(batchId)
+      }
+      true
+    } else {
+      // Return false as there is another writer.
+      false
+    }
+  }
+
+  /**
+   * Returns all file logs except the deleted files.
+   */
+  def allLogs(): Array[FileLog] = {
+    var latestId = getLatest().map(_._1).getOrElse(-1L)
+    while (true) {
+      if (latestId >= 0) {
+        val startId = getAllValidBatches(latestId, compactLength)(0)
+        try {
+          val logs = get(Some(startId), Some(latestId)).flatMap(_._2)
+          return compactLogs(logs).toArray
+        } catch {
+          case e: IOException =>
+            // Another process may delete the batch files when we are reading. However, it only
+            // happens when there is a compaction done. If so, we should retry to read the batches.
+            // Otherwise, this is a real IO issue and we should throw it.
+            val preLatestId = latestId
+            latestId = getLatest().map(_._1).getOrElse(-1L)
+            if (preLatestId == latestId) {
+              throw e
+            }
+        }
+      } else {
+        return Array.empty
+      }
+    }
+    Array.empty
+  }
+
+  /**
+   * Since all logs before `compactionBatchId` are compacted and written into the
+   * `compactionBatchId` log file, they can be removed. However, due to the eventual consistency of
+   * S3, the compaction file may not be seen by other processes at once. So we only delete files
+   * created `fileExpiredTimeMS` milliseconds ago.
+   */
+  private def deleteExpiredLog(compactionBatchId: Long): Unit = {
+    val expiredTime = System.currentTimeMillis() - fileExpiredTimeMS
+    fileManager.list(metadataPath, new PathFilter {
+      override def accept(path: Path): Boolean = {
+        try {
+          val batchId = getBatchIdFromFileName(path.getName)
+          batchId < compactionBatchId
+        } catch {
+          case _: NumberFormatException =>
+            false
+        }
+      }
+    }).foreach { f =>
+      if (f.getModificationTime <= expiredTime) {
```

Comment:
I made a mistake about `getModificationTime`. Only the directory's `getModificationTime` in S3 returns 0. File's getModificationTime returns the correct value. See https://issues.apache.org/jira/browse/HADOOP-12837

I also verified that in S3 by myself.